Docs 2.0 New Documentation. Batch 1 of content for review (#672)
graknlabs/docs: 0001-typedb/01-start/03-quickstart.md

---
pageTitle: Quickstart
keywords: getting started, typedb, typeql, tutorial, quickstart, overview
longTailKeywords: get started with typedb, typedb tutorial, typedb quickstart, learn typedb

> summary: Learn about the constructs of the TypeDB Schema, visualise a knowledge graph, perform read and write queries and explore the power of automated reasoning and analytics with TypeDB.

Reviewer: ```suggestion
summary: Learn about the constructs of the TypeDB Schema, visualize a knowledge graph, perform read and write queries and explore the power of automated reasoning and analytics with TypeDB.
```

toc: false
---

# Quickstart Guide

## Overview

> This quickstart goes through the step-by-step process of creating and interacting with a TypeDB database designed for an

Author: ```suggestion
This Quickstart guide goes through the step-by-step process of creating and interacting with a TypeDB database designed for an
```

[later](04-iam-schema.md).

<div class="note">
[Important]
We highly recommend completing this guide. Its goal is to prepare an environment for TypeDB exploration and development.
The resulting database schema and data will be needed for example queries throughout the documentation.
</div>

## Prerequisites

This quickstart takes advantage of TypeDB Studio, a desktop GUI tool for interacting with TypeDB. To proceed, please 
install both TypeDB and TypeDB Studio:

- [Install TypeDB](02-installation.md)
- [Install TypeDB Studio](../../02-clients/01-studio.md#get-typedb-studio)

IMPORTANT:
<div class="note">
[Important]
It’s recommended to use TypeDB and TypeDB Studio of the same version number in order to ensure compatibility.
</div>

## Initialize database

### Start TypeDB

Run the following command in a terminal:

```bash
typedb server
```

After showing the TypeDB ASCII logo and logging the bootup completion time, TypeDB is ready for connections.

![TypeDB server bootup message](../../images/quickstart/quickstart-launched-typedb.png)

### Connect to TypeDB

Launch TypeDB Studio, then:

1. Click the [**Connect to TypeDB**] button (right side of the toolbar)

> 2. Enter localhost:1729 in the [**Address**] field

Author: ```suggestion
2. Enter `localhost:1729` in the [**Address**] field
```

![Connection Manager Connected](../../images/studio/connection-interface-connected.png)

### Create a project

TypeDB Studio projects allow you to organize and save related queries for future reuse.

![Project Interface empty](../../images/quickstart/studio-projects-empty.png)

To create a new project:

1. Click the [**Open Project**] button in the [**Project**] panel (upper left)
2. Choose a directory for the project files
3. Click the [**Open**] button

The [**Project**] panel will now display the root directory and a nested hidden directory for unsaved files.

![Project Interface With Open Folder](../../images/studio/project-interface-created-folder.png)

### Create a database

1. Click the database icon to open the [**Manage Databases**] dialog (left side of the toolbar)

   ![Database Manager Empty](../../images/studio/studio-database.png)

2. Enter `iam` in the text field, and click the [**Create**] button next to it
3. Click the [**Close**] button in the bottom right

   ![Database Manager With Phone Calls Database](../../images/studio/databases-interface-iam-database.png)

4. Select `iam` from the database dropdown (next to the database icon)

### Prepare a TQL file

> A [TypeDB schema](../02-dev/02-schema.md) contains entity, relation, and attribute type definitions that make up the 

Reviewer: ```suggestion
A [TypeDB schema](../02-dev/02-schema.md) contains entity, relation and attribute type definitions that make up the 
```

Author: Isn't it an Oxford comma? I believe it's almost mandatory in American English.

        };
    } then {
        $p has validity true;
    };
    
rule add-view-access:
    when {
        $modify isa action, has action-name "modify_file";
        $view isa action, has action-name "view_file";
        $ac_modify (accessed-object: $obj, valid-action: $modify) isa access;
        $ac_view (accessed-object: $obj, valid-action: $view) isa access;
        (permitted-subject: $subj, permitted-access: $ac_modify) isa permission;
    } then {
        (permitted-subject: $subj, permitted-access: $ac_view) isa permission;
    };
```

### Upload the schema

To execute the TypeQL statements in the opened file and send them as queries:

1. Ensure the [**Session type**] (schema / data) switch (next to the database dropdown) is set to `schema`
2. Ensure the [**Transaction type**] (write / read) switch is set to `write`

   ![Select transaction type](../../images/quickstart/studio-select-transaction-type.png)

3. Click the green “play” button (middle of the toolbar) to start the transaction
4. Click the “checkmark” button (left of the “play” button) to commit the changes

The transaction has been committed, and `iam` database now has a schema.

> The Types panel will now display the entity, relation, and attribute types within a type hierarchy of the schema.

Reviewer: ```suggestion
The Types panel will now display the entity, relation and attribute types within a type hierarchy of the schema.
```

![Types browser with IAM schema](../../images/quickstart/studio-types-browser-iam.png)

Data can now be inserted.

## First important queries

### Read the schema

TypeQL can be used to query the schema.

To execute a simple schema query:

1. Ensure the [**Session type**] (schema / data) switch is set to `schema` (next to the database dropdown)
2. Ensure the [**Transaction type**] (write / read) switch is set to `read`
3. Click the `+` icon right from the [**Project**] panel next to the `iam-schema.tql` tab in the Text-editor panel
4. Copy the TypeQL statement below
5. Click the green “play” button

```typeql
match $s sub subject;
```

The above query returns the subject type and its subtypes and displays the results as a graph.

![IAM schema graph](../../images/quickstart/studio-iam-schema.png)

### Insert data

> We will insert data the same way we created the schema, by creating a TQL file in our project and executing it.

Reviewer: ```suggestion
We will insert data the same way we created the schema, by creating a `.tql` file in our project and executing it.
```

insert $p (permitted-subject: $s, permitted-access: $ac) isa permission;

match $s isa subject, has full-name "Pearle Goodman"; $o isa object, has path "README.md"; 
      $a isa action, has action-name "modify_file"; $ac (accessed-object: $o, valid-action: $a) isa access; 
insert $p (permitted-subject: $s, permitted-access: $ac) isa permission;
```

To execute the TypeQL statements copied from code block above:

1. Ensure the [Session type] (schema / data) switch (next to the database dropdown) is set to `data`
2. Ensure the [Transaction type] (write / read) switch is set to `write`
3. Click the green “play” button
4. Click the “checkmark” button

The transaction has been committed, and data can now be queried.

### Read data

Below is a simple query that returns all person and operation entities (and their names) as well as all file entities 
(and their paths).

To execute a simple data query:

1. Click the `+` icon right from the [**Project**] panel next to the `iam-data.tql` tab in the Text-editor panel
2. Ensure the [Session type] (schema / data) switch (next to the database dropdown) is set to `data` 
3. Ensure the [Transaction type] (write / read) switch is set to `read`

> 4. Replace the TypeQL statement in the Text-editor panel with the one below

Reviewer: ```suggestion
4. Replace the TypeQL statement in the text editor panel with the one below
```

5. Click the green “play” button

```typeql
match $p isa person, has full-name $fn; $f isa file, has path $fp;
```

> The above query returns all person entities (instances of person type) with their names as well as all file entities 

Reviewer: ```suggestion
The above query returns all `person` entities (instances of `person` type) with their `full-name` attributes as well as all `file` entities 
```

> with their paths and displays the results as a graph.

Reviewer: ```suggestion
with their `path` attributes and displays the results as a graph.
```

<!-- test-ignore -->
```typeql
#
# Copyright (C) 2022 Vaticle
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#

define

company sub entity,
    owns name,
    plays company-membership:parent-company;

company-membership sub relation,
    relates parent-company,
    relates company-member;

parent-company sub attribute,
    value string;

#rule attribute-parent-company:
#    when {
#        (parent-company: $c, company-member: $t) isa company-membership;
#        $c has name $c-name;
#        ?name-value = $c-name
#    } then {
#        $t has parent-company ?name-value;

> #    };

Reviewer 2: Do we want to include this commented out rule?
